src/data_source/storage/sql.rs

```diff
@@ -1170,6 +1188,518 @@ where
+        // Get all the nodes in the path to the index.
+        // Order by pos DESC is to return nodes from the leaf to the root
+        let nodes = self
+            .query(
+                &format!(
+                    "SELECT DISTINCT ON (pos) *
+                    FROM {state_type}
+                    WHERE pos @> $1 and created <= $2
+                    ORDER BY pos DESC , created DESC;"
+                ),
+                [sql_param(&LTree::from(path.iter())), sql_param(&created)],
+            )
+            .await?;
+
+        let nodes: Vec<_> = nodes.map(|res| Node::try_from(res?)).try_collect().await?;
+        // insert all the hash ids to a hashset which is used to query later
+        // HashSet is used to avoid duplicates
+        let mut hash_ids = HashSet::new();
+        nodes.iter().for_each(|n| {
+            hash_ids.insert(n.hash_id);
+            if let Some(children) = &n.children {
+                hash_ids.extend(children);
+            }
+        });
+
+        // Find all the hash values and create a hashmap
+        // Hashmap will be used to get the hash value of the nodes children and the node itself.
+        let hashes_query = self
+            .query(
+                "SELECT * FROM hash WHERE id = ANY( $1)",
+                [sql_param(&hash_ids.into_iter().collect::<Vec<i32>>())],
+            )
+            .await?;
+        let hashes: HashMap<_, _> = hashes_query
+            .map(|row| HashTableRow::try_from(row?).map(|h| (h.id, h.value)))
+            .try_collect()
+            .await?;
+
+        // The merkle tree commitment from the header for the snapshot should match the root node hash in the table
+        // The last node in nodes vector is the root node
+        // Merkle tree commitment is calculated using the children
+        let root_node = nodes.last().ok_or(QueryError::Error {
+            message: "No node found in the db".to_owned(),
+        })?;
+        let digest = if let (Some(children), Some(bitvec)) =
+            (&root_node.children, &root_node.children_bitvec)
+        {
+            let mut next_child = 0;
+            let root_children_values = bitvec
+                .iter()
+                .map(|b| {
+                    if b {
+                        let child = children[next_child];
+                        next_child += 1;
+                        let hash = hashes.get(&child).ok_or(QueryError::NotFound)?;
+                        State::T::deserialize_compressed(hash.as_slice()).map_err(|e| {
+                            QueryError::Error {
+                                message: format!("failed to deserialize hash to Nodevalue err {e}"),
+                            }
+                        })
+                    } else {
+                        Ok(State::T::default())
+                    }
+                })
+                .collect::<QueryResult<Vec<_>>>()?;
+
+            State::Digest::digest(&root_children_values).map_err(|e| QueryError::Error {
```

Comment:
Ayy this is annoying. I think this check is still not quite right (might give false positives): because we are recomputing the root hash based on the hashes of children saved in the database, the check will pass as long as we have the correct version of the immediate children of the root, even if nodes farther down the path are missing their latest version.

I think what we actually need to check is that the entire path is a valid merkle path and the value of the root node is consistent with the header field. So basically traverse each node in the path, recompute it's digest based on its children's digests and compare it to the value field of the node. Then finally check the root against `merkle_commitment`.

Reply:
I think we are also missing a test for the case where this check fails